### PR TITLE
perf: Reuse segment reductions for bincount

### PR DESCRIPTION
**Pull request by n-gao**
Add a compiled `bincount` backed by `segment_sum` and use it at every JAX counting call site. Single-bin counts use a reduction; clipping, padding, empty inputs and integer dtypes retain their existing behavior.

N₂ GCMC attempted moves/s on CPU core 1; float64, median of six interleaved 512-cycle windows. Initialization and compilation excluded.

| Before this PR | With this PR | Change |
|---:|---:|---:|
| 2,227 | 2,335 | +4.8% |

Validation: 254 CPU and 176 GPU regression tests passed; Ruff and source typing passed.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#334** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/334?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#337** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/337?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#336** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/336?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/336?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#335** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#333** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/333?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#332** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#331** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#330** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/330?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#329** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#328** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/328?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by felixmusil** [comment]
Amended (with Felix's go-ahead) to unblock the `lowest-direct` CI matrix: `jax.enable_x64` is only public from jax 0.8.0, and `pyproject` allows `jax>=0.7.2`, so `TestBincount::test_matches_jax` raised `AttributeError` under jax 0.7.2. Replaced it with a local `_x64` context manager over `jax.config.update`, matching this module's `_disable_x64` fixture. Verified `test/core/test_segment.py`: 113 passed under both `--resolution lowest-direct` (jax 0.7.2) and `highest` (0.11.2). Also restacked #336/#337/#334 onto current main (#331–#335 merged) and squashed the empty leftover commit on #334.

**Review by felixmusil — APPROVED** [review]

**Comment by felixmusil** [comment]
### Merge activity

* **Sep 18, 1:45 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/336).
* **Sep 18, 1:45 PM UTC**: @felixmusil merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/336).
